src/Symfony/Bundle/SecurityBundle/Security/FirewallContext.php

```diff
@@ -23,11 +23,25 @@ class FirewallContext
 {
     private $listeners;
     private $exceptionListener;
+    private $config;
 
-    public function __construct(array $listeners, ExceptionListener $exceptionListener = null)
+    public function __construct(array $listeners, ExceptionListener $exceptionListener = null, FirewallConfig $config = null)
     {
         $this->listeners = $listeners;
         $this->exceptionListener = $exceptionListener;
+
+        if (null === $config) {
+            @trigger_error(sprintf('"%s()" expects an instance of "%s" as third argument since version 3.2 and will throw an exception in 4.0 if not provided.', __METHOD__, FirewallConfig::class), E_USER_DEPRECATED);
```

Comment:
The thing is that the firewall is not depending on this config class for being configured, it can simply live without it. This "aware" interface is more about sharing by exposing, and I guess it was not designed this way before because the less you expose in security the more it's safe, you've noticed that yourself with the listeners config. So this class does not really add any value unless it comes to be collected and displayed (i.e in the profiler).

Also, by removing `null` as default you should expect a php error instead of an exception unless you throw it yourself in 4.0, meaning keeping the `if` too to perform the check.

So actually the message should be something like `"%s()" expects an instance of "%s" as third argument since version 3.2 and will trigger an error in 4.0`.